text/0000-ember-data-improved-ajax.md

```diff
@@ -0,0 +1,207 @@
+- Start Date: 2016-10-11
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Add public hooks on `rest`-adapter which can be used to customize the
+properties of an AJAX request being made for an adapter operation.
+
+1) hooks to get properties for a request which should be made
+  - `methodForRequest`
+  - `headersForRequest`
+  - `urlForRequest`
+    - `host`
+    - `namespace`
+    - `pathForRequest`
+    - `queryForRequest`
+  - `bodyForRequest`
+
+2) hook for making the request
+  - `makeRequest`
+
+# Motivation
+
+Currently `host`, `namespace`, `urlForXXX()` and `headers` is the only public
+API on the `rest`-adapter which allow you to customize the request for an
+adapter operation like `findRecord`, `createRecord`, .... Technically `ajax`
+and `ajaxOptions` are private, but they have been overwritten in apps and
+addons due to the lack of a public API.
+
+The aim of the `ds-improved-ajax` feature implemented in
+[#3099](https://github.com/emberjs/data/pull/3099) was to address the lack of
+public API for customizing a request. Unfortunately, it hasn't gone through a
+proper RFC process :pensive:. The feature is not enabled and available in a
+`release`, so this RFC should flesh out the specific API for the feature before
+it can be enabled eventually.
+
+The result of this RFC should make sure that the new hooks allow developers to
+customize a request with only public API.
+
+# Detailed design
+
+An adapter operation is split up into 2 steps
+
+- 1) collecting the properties for the request, i.e. `_requestFor`
+- 2) executing the request, i.e. `makeRequest`
+
+So for a `findRecord` it would look like this:
+
+``` js
+findRecord(store, type, id, snapshot) {
+  let requestOptions = this._requestFor({
+    store, type, id, snapshot,
+    requestType: "findRecord"
+  });
+
+  return RSVP.resolve(requestOptions).then((options) => this.makeRequest(options) );
+}
+```
+
+### `_requestFor`
+
+Each hook for step 1) is invoked with all the arguments which are passed to the
+corresponding adapter method. Additionally `requestType` is set so the hooks
+know for which specific adapter operation they are invoked. To allow for
+asynchronous configurations (e.g. pre-flight request to refresh an
+authentication token), `_requestFor` returns a promise which resolves with an
+object describing the request to make. The `_requestFor` method is private and
+calls the new public hooks.
+
+``` js
+_requestFor(params) {
+  let method  = this.methodForRequest(params);
+  let url     = this.urlForRequest(params);
+  let headers = this.headersForRequest(params);
+  let body    = this.bodyForRequest(params);
+
+  return RSVP.hash({ method, url, headers, body });
```

Comment:
Should `_requestFor` be made a public method in case the user wants to add some form of config other then method, url, headers or body?